Add pending bodies for async values

smfoote opened the pull request:
Full testing of this feature will require a better testing framework,
but basic testing is included here, as well as manual testing using the
sandbox.

smfoote committed:
Add pending bodies for async values

Full testing of this feature will require a better testing framework,
but basic testing is included here, as well as manual testing using the
sandbox.

smfoote commented:
Fixes #45

smfoote commented:
Example:

```
{#asyncResults}
   {>result/}
{:pending}
  Results are loading <span class="loading-spinner"></span>
{/asyncResults}
```

jimmyhchan commented on src/runtime.js:
```diff
@@ -151,6 +151,7 @@ let tornado = {
    */
   exists(val, placeholderNode, bodies, context) {
     if (this.util.isPromise(val)) {
+      placeholderNode = this.insertPendingBody(placeholderNode, bodies.pending, context) || placeholderNode;
```
does this need to live in runtime? Is there a general (DRYer) way to handle `{:else}` and `{:pending}` etc. If we add a few more bodies, we will have to litter these everywhere.

smfoote replied:
Yeah... I don't know if there is a DRYer way, but I think that `pending` needs to live in the runtime because we don't know until runtime if the thing is a promise.

prashn64 replied:
You can do something like this:

```
placeHolderNode = this.promiseCreator(placeholderNode, bodies.pending, context, function() {
  //custom code for this particular function(section, exists, etc.)
});



promiseCreator = function(val, placeholderNode, bodies.pending, context, cb) {
   placeholderNode = this.insertPendingBody(placeholderNode, bodies.pending, context) || placeholderNode;

  val.then(function (data) {
    cb();
  });
}
```

smfoote replied:
I don't think that solves the problem of what happens when we add a new "standardized" body.

prashn64 replied:
It would not, but it does solve the current clutter problem at least.  We could eventually make a 'nonPromiseCreator' function as well that handles nonasync bodies

smfoote replied:
Okay, this all may be premature optimization. I don't think there are any other insertion points in the lifecycle, so adding more "standardized" bodies probably won't be a problem.

Having said that, I suppose I can imagine a `reject` or `fail` body to replace `else` when a promise rejects. But, that would be in a completely different part of the lifecycle.

Do you think this needs to be resolved before we pull this PR?

prashn64 replied:
since it affects more than just the pending we can address it separately.

jimmyhchan replied:
Another idea, should this live as a helper instead?  the issue with original Dust is that it never had a good way of standardizing bodies. i'd like this explicit with lifecycles before we start teh slippery slope arguments.

smfoote replied:
What would the helper look like?

jimmyhchan replied:
```
{@existsWithPromise key=data}
   show when done
{:pending}
  <div class="pending"></div>
{/existsWithPromise}
```

It'll be the same logic. What if any runtime level feature be a helper and every compile level feature is in the grammar/compiler.

smfoote replied:
I don't know if I like the "everything is a helper" idea. Why should we have `exists` if we have `existsWithPromise`? What if `data` is only sometimes a promise? How are developers going to remember all of these helpers? It seems that this solution is good for the project maintainers at the expense of the project users.

jimmyhchan replied:
what makes `eq` worthy of being a helper?   Why does `?exists` not live as a helper? IMO, having `?` look at the data makes the distinction muddy. 

for more muddiness why not have it as a filter?

`{#reddit.dat  | pending}`

smfoote replied:
Alright, I see what you mean. So, should we get rid of `exists` and `section`? I think there is value in making the most common use cases a part of the runtime, allowing them to be optimized beyond what is possible in a helper.

jimmyhchan replied:
agree we should optimize the common usecase. but as an implementation `exists` and `section` should be no different then any other helper. We should build them on top of the same APIs helper authors use. 

It'll be nice to have a few APIs for the common use cases:
- extend built-ins
- work like how filters are being used today
- create new DOM nodes
- modify a copy of context.

etc.

smfoote replied:
I really like the idea of coming up with a spec for helpers. We should put that in the Constitution.

prashn64 commented on src/runtime.js:
```diff
@@ -359,6 +370,27 @@ let tornado = {
+  insertPendingBody(placeholderNode, body, context) {
+    if (body) {
+      let div = document.createElement('div');
+      div.setAttribute('class', 'pending');
```
I'm suspicious of adding extra decoration like this.  Is this necessary or just a nice way to see a pending html element?

smfoote replied:
This is necessary, unless we want to add some other kind of ugly decoration. Here is the problem that the `div.pending` wrapper solves:

Fragment before inserting `pending` body:

```
My content will go here: [single empty string text node, acting as an insertion point]
```

Fragment after inserting pending body:

```
My content will go here: While you wait, here's a <b>really</b> funny joke. Why did the rubber band ...
```

Now, where we had a single, empty-string text node we have 3 sibling nodes: text, <b>, and text. When my content arrives, we need to insert it, but where? Our insertion point is gone, replaced by 3 text nodes.

In theory, we could count how many top level nodes are in a given pending body, and then replace that many nodes when the real content arrives, but that is a lot of overhead, and quite messy.

So, instead, we wrap the pending body in a div (with a convenient class name), and we maintain the single insertion point, and we document the fact that a pending body gets a wrapper. I think this div wrapper can be viewed as a feature rather than a nuisance. An empty pending body will give you div in the right place, where you can put a loading spinner, or whatever you want using CSS.

prashn64 replied:
Do you forsee bugs where someone happened to have a div with a pending class or does that case just never come into play?

smfoote replied:
We could change the class to `tornado-pending`, or add the ability to customize that class name, but I would wait for a bug report before we do either.

prashn64 replied:
I do recomend tornado-pending

prashn64 commented:
Looks good other than the couple of line note comments

smfoote commented:
Is it reasonable to merge this (after rebasing and addressing some comments above), then work out how to make Tornado built-ins use the helper API? The external API won't change.

jimmyhchan commented:
this is a new feature so it seems this can go in at any time.  We are still at 0.X so it's fine to break APIs and take it out before 1.x if we really need to.
